// gluestack-ui/dist/util/config/next-config-helper.js
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "path", "fast-glob", "fs-extra", "../../config", ".", "./config-types", "path", "..", "../init", "os"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.generateConfigNextApp = generateConfigNextApp;
    const path = __importStar(require("path"));
    const fast_glob_1 = __importDefault(require("fast-glob"));
    const fs_extra_1 = require("fs-extra");
    const config_1 = require("../../config");
    const _1 = require(".");
    const config_types_1 = require("./config-types");
    const path_1 = require("path");
    const __1 = require("..");
    const init_1 = require("../init");
    const os_1 = __importDefault(require("os"));
    const _homeDir = os_1.default.homedir();
    //next project type initialization
    function getNextProjectType(cwd) {
        return __awaiter(this, void 0, void 0, function* () {
            const files = yield fast_glob_1.default.glob('**/*', {
                cwd,
                deep: 3,
                ignore: config_types_1.PROJECT_SHARED_IGNORE,
            });
            const isNextProject = files.find((file) => file.startsWith('next.config.'));
            if (!isNextProject) {
                return undefined;
            }
            const isUsingSrcDir = yield (0, fs_extra_1.pathExists)(path.resolve(cwd, 'src'));
            const isUsingAppDir = yield (0, fs_extra_1.pathExists)(path.resolve(cwd, `${isUsingSrcDir ? 'src/' : ''}app`));
            if (isUsingAppDir) {
                return isUsingSrcDir ? 'next-app-src' : 'next-app';
            }
            return isUsingSrcDir ? 'next-pages-src' : 'next-pages';
        });
    }
    function resolvedNextJsPaths(resultConfig) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c;
            const resolvedNextJsPaths = {
                tailwind: {
                    config: (0, _1.pathResolver)(resultConfig.tailwind.config),
                    css: (0, _1.pathResolver)(resultConfig.tailwind.css),
                },
                config: {
                    postCssConfig: (0, _1.pathResolver)(resultConfig.config.postCssConfig || ''),
                    nextConfig: (0, _1.pathResolver)(resultConfig.config.nextConfig || ''),
                    tsConfig: (0, _1.pathResolver)(resultConfig.config.tsConfig || ''),
                },
                app: {
                    entry: (0, _1.pathResolver)(resultConfig.app.entry || ''),
                    type: (_a = resultConfig === null || resultConfig === void 0 ? void 0 : resultConfig.app) === null || _a === void 0 ? void 0 : _a.type,
                    registry: ((_b = resultConfig === null || resultConfig === void 0 ? void 0 : resultConfig.app) === null || _b === void 0 ? void 0 : _b.registry)
                        ? resultConfig.app.registry.replace(/\\/g, '/')
                        : undefined,
                    page: ((_c = resultConfig === null || resultConfig === void 0 ? void 0 : resultConfig.app) === null || _c === void 0 ? void 0 : _c.page)
                        ? path.resolve(_1._currDir, resultConfig.app.page)
                        : '',
                },
            };
            return resolvedNextJsPaths;
        });
    }
    //project specific initialization: nextjs
    function initNatiwindNextApp(resolvedConfig, permission, isNextjs15) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            try {
                if (((_b = (_a = resolvedConfig.app) === null || _a === void 0 ? void 0 : _a.entry) === null || _b === void 0 ? void 0 : _b.includes('layout')) &&
                    resolvedConfig.app.registry) {
                    // if app router add registry file to root
                    const registryPath = isNextjs15 ? ['nextjs', 'next15'] : ['common'];
                    const templatesPath = (0, path_1.join)(_homeDir, config_1.config.gluestackDir, config_1.config.templatesDir);
                    const registryContent = yield (0, fs_extra_1.readFile)((0, path_1.join)(templatesPath, ...registryPath, 'registry.tsx'), 'utf8');
                    yield (0, fs_extra_1.writeFile)(resolvedConfig.app.registry, registryContent, 'utf8');
                }
                yield (0, init_1.commonInitialization)(config_1.config.nextJsProject, resolvedConfig, permission);
            }
            catch (err) {
                throw new Error(err.message);
            }
        });
    }
    function generateConfigNextApp(permission, isNextjs15) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const projectType = yield getNextProjectType(_1._currDir);
            const entryPath = yield (0, _1.getFilePath)(['**/*layout.*', '**/*_app.*']);
            const globalCssPath = yield (0, _1.getFilePath)([
                '**/*globals.css',
                '**/*global.css',
            ]);
            const tailwindConfigPath = yield (0, _1.getFilePath)(['tailwind.config.*']);
            const postCssConfigPath = yield (0, _1.getFilePath)(['postcss.config.*']);
            const nextConfigPath = yield (0, _1.getFilePath)(['next.config.*']);
            const tsConfigPath = yield (0, _1.getFilePath)(['tsconfig.*']);
            let registryPath = '';
            if (projectType === null || projectType === void 0 ? void 0 : projectType.includes('app')) {
                const appDirectory = (0, _1.findDirectory)(_1._currDir, ['src/app', 'app']);
                registryPath = path.resolve(_1._currDir, appDirectory, 'registry.tsx');
            }
            const pagePath = entryPath.includes('layout.')
                ? yield (0, _1.getFilePath)(['**/*page.*'])
                : undefined;
            const gluestackConfig = {
                tailwind: {
                    config: tailwindConfigPath.length
                        ? tailwindConfigPath
                        : 'tailwind.config.js',
                    css: globalCssPath.length ? globalCssPath : 'globals.css',
                },
                app: {
                    entry: entryPath,
                    // write a function to get current components path
                    components: config_1.config.writableComponentsPath,
                },
            };
            const resolvedGluestackConfig = {
                tailwind: {
                    config: tailwindConfigPath.length
                        ? tailwindConfigPath
                        : 'tailwind.config.js',
                    css: globalCssPath.length ? globalCssPath : 'globals.css',
                },
                config: {
                    postCssConfig: postCssConfigPath.length
                        ? postCssConfigPath
                        : 'postcss.config.js',
                    nextConfig: nextConfigPath.length ? nextConfigPath : 'next.config.js',
                    tsConfig: tsConfigPath.length ? tsConfigPath : 'tsconfig.json',
                },
                app: {
                    type: projectType,
                    entry: entryPath,
                    registry: registryPath,
                    page: pagePath,
                },
            };
            // generateConfig(gluestackConfig);
            const resolvedConfig = yield resolvedNextJsPaths(resolvedGluestackConfig);
            const filesTobeEnsured = [
                (_a = resolvedConfig.app.registry) !== null && _a !== void 0 ? _a : '',
                resolvedConfig.config.tsConfig,
                resolvedConfig.tailwind.css,
                resolvedConfig.config.postCssConfig,
                (0, _1.pathResolver)('nativewind-env.d.ts'),
            ];
            const filesEnsured = yield (0, __1.ensureFilesPromise)(filesTobeEnsured);
            if (permission && filesEnsured) {
                yield yield initNatiwindNextApp(resolvedConfig, permission, isNextjs15);
            }
        });
    }
});
