// gluestack-ui/dist/config.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.config = void 0;
    const config = {
        repoUrl: 'https://github.com/gluestack/gluestack-ui.git',
        gluestackDir: '.gluestack/cache/gluestack-ui',
        componentsResourcePath: 'src/components/ui',
        nativeWindRootPath: 'nativewind',
        expoProject: 'expo',
        nextJsProject: 'nextjs',
        reactNativeCLIProject: 'react-native-cli',
        tailwindConfigRootPath: 'src/gluestack-ui/templates/tailwind.config.js',
        writableComponentsPath: 'components/ui',
        branchName: 'main', // branch name for the gluestack-ui repo to be used for the CLI
        style: 'nativewind',
        providerComponent: 'gluestack-ui-provider',
        gluestackUIPattern: '@/components/ui/',
        templatesDir: 'packages/gluestack-ui/templates',
        packageManager: null,
        ignoreComponents: ['utils'],
    };
    exports.config = config;
});
